bundles/org.openhab.binding.solarforecast/src/main/java/org/openhab/binding/solarforecast/internal/forecastsolar/ForecastSolarObject.java

```diff
@@ -0,0 +1,321 @@
+                    } catch (DateTimeParseException dtpe) {
+                        logger.warn("Exception parsing time {} Reason: {}", dateStr, dtpe.getMessage());
+                        return;
+                    }
+                }
+            } catch (JSONException je) {
+                logger.debug("Error parsing JSON response {} - {}", content, je.getMessage());
+            }
+        }
+    }
+
+    public boolean isValid() {
+        return !(wattHourMap.isEmpty() || wattMap.isEmpty());
+    }
+
+    public boolean isExpired() {
+        return expirationDateTime.isBefore(Instant.now());
+    }
+
+    public double getActualEnergyValue(ZonedDateTime queryDateTime) {
+        if (!isValid()) {
+            return -1;
+        }
+        Entry<ZonedDateTime, Double> f = wattHourMap.floorEntry(queryDateTime);
+        Entry<ZonedDateTime, Double> c = wattHourMap.ceilingEntry(queryDateTime);
+        if (f != null && c == null) {
+            // only floor available
+            if (f.getKey().toLocalDate().equals(queryDateTime.toLocalDate())) {
+                // floor has valid date
+                return f.getValue() / 1000.0;
+            } else {
+                // floor date doesn't fit
+                return -1;
+            }
+        } else if (f == null && c != null) {
+            if (c.getKey().toLocalDate().equals(queryDateTime.toLocalDate())) {
+                // only ceiling from correct date available - no valid data reached yet
+                return 0;
+            } else {
+                // ceiling date doesn't fit
+                return -1;
+            }
+        } else if (f != null && c != null) {
+            // ceiling and floor available
+            if (f.getKey().toLocalDate().equals(queryDateTime.toLocalDate())) {
+                if (c.getKey().toLocalDate().equals(queryDateTime.toLocalDate())) {
+                    // we're during suntime!
+                    double production = c.getValue() - f.getValue();
+                    long floorToCeilingDuration = Duration.between(f.getKey(), c.getKey()).toMinutes();
+                    if (floorToCeilingDuration == 0) {
+                        return f.getValue() / 1000.0;
+                    }
+                    long floorToQueryDuration = Duration.between(f.getKey(), queryDateTime).toMinutes();
+                    double interpolation = (double) floorToQueryDuration / (double) floorToCeilingDuration;
+                    double interpolationProduction = production * interpolation;
+                    double actualProduction = f.getValue() + interpolationProduction;
+                    return actualProduction / 1000.0;
+                } else {
+                    // ceiling from wrong date, but floor is valid
+                    return f.getValue() / 1000.0;
+                }
+            } else {
+                // floor invalid - ceiling not reached
+                return 0;
```

Comment:
Same here, error or is 0 a valid result in this case?

Reply:
Same as above - floor is from day before, ceiling is valid => sunrise not reached yet, valid zero